File: 014_Gas Station.cpp
/*
Gas Station

Given two arrays gas and cost, return the starting gas station index if you can travel around the circuit once, otherwise -1.

Strategy: If total gas >= total cost, there is a solution. Greedy: accumulate tank and reset start when tank < 0.
*/

#include <vector>
#include <iostream>
using namespace std;

int canCompleteCircuit(const vector<int>& gas, const vector<int>& cost) {
    int n = gas.size();
    int total = 0, tank = 0, start = 0;
    for (int i = 0; i < n; ++i) {
        int diff = gas[i] - cost[i];
        total += diff;
        tank += diff;
        if (tank < 0) {
            start = i + 1;
            tank = 0;
        }
    }
    return total >= 0 ? start : -1;
}

int main() {
    cout << "GasStation 1: " << canCompleteCircuit({1,2,3,4,5}, {3,4,5,1,2}) << '\n'; // 3
    cout << "GasStation 2: " << canCompleteCircuit({2,3,4}, {3,4,3}) << '\n'; // -1
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(1)
*/

File: 014_Gas Station.py
"""
Gas Station

Return starting gas station index to complete circuit, or -1 if impossible.
"""

from typing import List

def can_complete_circuit(gas: List[int], cost: List[int]) -> int:
    total = tank = start = 0
    for i in range(len(gas)):
        diff = gas[i] - cost[i]
        total += diff
        tank += diff
        if tank < 0:
            start = i + 1
            tank = 0
    return start if total >= 0 else -1


if __name__ == '__main__':
    print('GasStation 1:', can_complete_circuit([1,2,3,4,5], [3,4,5,1,2]))
    print('GasStation 2:', can_complete_circuit([2,3,4], [3,4,3]))

# Time Complexity: O(n)
# Space Complexity: O(1)
